core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1108,11 +1118,14 @@ private[spark] class BlockManager(
   def removeBlock(blockId: BlockId, tellMaster: Boolean = true): Unit = {
     logDebug(s"Removing block $blockId")
     val info = blockInfo.get(blockId).orNull
-    if (info != null) {
+    if (info != null && !pendingToRemove.containsKey(blockId)) {
+      pendingToRemove.put(blockId, currentTaskAttemptId)
```

Comment:
Actually, on closer inspection I believe this is still not safe. You need to use something like `pendingToRemove.putIfAbsent` to do this atomically.